From 12457472b37ac8b0b7c2e5a61b91cdceb747b6d9 Mon Sep 17 00:00:00 2001
From: Liu Ying <victor.liu@nxp.com>
Date: Fri, 13 Oct 2017 12:33:39 +0800
Subject: [PATCH 2643/5242] MLK-16581-4 gpu: imx: framegen: Get pll & pixel
 clock rates before setting their rates

commit  69323b9f418b0473ac340173058fa982725e0e20 from
https://source.codeaurora.org/external/imx/linux-imx.git

Due to i.MX8 clock issue, we need to get pll and pixel clock rates
before setting their rates when system resumes back from PM sleep mode,
otherwise, we'll fail to set the clock rates.  So, this is a workaround
and it can be removed when the clock issue is properly fixed.

Signed-off-by: Liu Ying <victor.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/dpu/dpu-framegen.c |    9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/gpu/imx/dpu/dpu-framegen.c b/drivers/gpu/imx/dpu/dpu-framegen.c
index e59b8bf..b068e7c 100644
--- a/drivers/gpu/imx/dpu/dpu-framegen.c
+++ b/drivers/gpu/imx/dpu/dpu-framegen.c
@@ -279,6 +279,15 @@ void framegen_cfg_videomode(struct dpu_framegen *fg, struct drm_display_mode *m)
 			pll_clock_rate = disp_clock_rate * 8;
 	}
 
+	/*
+	 * To workaround setting clock rate failure issue
+	 * when the system resumes back from PM sleep mode,
+	 * we need to get the clock rates before setting
+	 * their rates, otherwise, setting the clock rates
+	 * will fail.
+	 */
+	clk_get_rate(fg->clk_pll);
+	clk_get_rate(fg->clk_disp);
 	clk_set_rate(fg->clk_pll, pll_clock_rate);
 	clk_set_rate(fg->clk_disp, disp_clock_rate);
 }
-- 
1.7.9.5

